build: Add Boost.Process usage check

hebasto opened the pull request:
This PR adds a check that Boost.Process can be used without linking any libraries (header-only).
Disable the functionality if that is not the case.

Fixes bitcoin/bitcoin#24314.

hebasto committed:
build: Add Boost.Process usage check

Check that Boost.Process can be used without linking any libraries
(header-only). Disable the functionality if that is not the case.

prusnak reviewed (approved):
Approach ACK 7b80c36

MarcoFalke commented:
> On master (https://github.com/bitcoin/bitcoin/commit/3ace3a17c9bce606cea05192f0da3ac62ac69dda), building on Bionic with system packages is https://github.com/bitcoin/bitcoin/pull/20744#issuecomment-1028979276.

The same happens on CentOS 8 with system packages.

I presume a workaround would be to disable external signer?

hebasto commented:
@MarcoFalke 
> I presume a workaround would be to disable external signer?

Correct. If `--enable-external-signer` is provided explicitly, `configure` fires an error.

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
49f0cc056e43adb607f6e930d64042a309622f60b69e26ac04ac4e2bd82e2774  guix-build-7b80c36c6f27/output/aarch64-linux-gnu/SHA256SUMS.part
a89e14d723bad20dc01bf7069590cbfc192436cf0807dcaa888a973c64d55fd3  guix-build-7b80c36c6f27/output/aarch64-linux-gnu/bitcoin-7b80c36c6f27-aarch64-linux-gnu-debug.tar.gz
214fcc64a06de3bce3d7dfe375f6049d9df406dc9cce2d0c16f35936d33eaea0  guix-build-7b80c36c6f27/output/aarch64-linux-gnu/bitcoin-7b80c36c6f27-aarch64-linux-gnu.tar.gz
f6008cf4266bad983850e6d015f389973772b585cd255b61b1e3f5b0d5b968e2  guix-build-7b80c36c6f27/output/arm-linux-gnueabihf/SHA256SUMS.part
9f3c8fc76e1ba17425edc3e677519a999bb4875b53c481a22ee6f1b32473db63  guix-build-7b80c36c6f27/output/arm-linux-gnueabihf/bitcoin-7b80c36c6f27-arm-linux-gnueabihf-debug.tar.gz
52f01fecd6f0dd4816aec7b6ea29fc191aa500c621a563ca1b53d1dd233075d8  guix-build-7b80c36c6f27/output/arm-linux-gnueabihf/bitcoin-7b80c36c6f27-arm-linux-gnueabihf.tar.gz
af2563e77173136f784b1641324b8a64d37bd3d9913979104ffb009f24c320b0  guix-build-7b80c36c6f27/output/arm64-apple-darwin/SHA256SUMS.part
968b328ebb89dda8b06fb098f3e3a926d2d9bf9908860341790dea2a75de4b54  guix-build-7b80c36c6f27/output/arm64-apple-darwin/bitcoin-7b80c36c6f27-arm64-apple-darwin.tar.gz
3e84b816e49867545ffb1e04761d8fc415c1ccc00ae90bad4e58afb9df555410  guix-build-7b80c36c6f27/output/arm64-apple-darwin/bitcoin-7b80c36c6f27-osx-unsigned.dmg
a9057ab1d0fee08e03c0dbf03688e5a68f159e427ae79f608ccd892386a0ae19  guix-build-7b80c36c6f27/output/arm64-apple-darwin/bitcoin-7b80c36c6f27-osx-unsigned.tar.gz
e590331ed328a49f4b60d18df6127be23629a37f1d5e793797391d2e7c57c718  guix-build-7b80c36c6f27/output/dist-archive/bitcoin-7b80c36c6f27.tar.gz
43252a3ca684d3743cd7b32ae9d7517a9c0ef543d525a9c6e43a8313ea3bde6d  guix-build-7b80c36c6f27/output/powerpc64-linux-gnu/SHA256SUMS.part
9b7f78c75f564b849dc106d844242f48ad961fb98147334843197067299199a1  guix-build-7b80c36c6f27/output/powerpc64-linux-gnu/bitcoin-7b80c36c6f27-powerpc64-linux-gnu-debug.tar.gz
457ae9e039fe7fd226de9b1ac0cb0496a22678ef733c0e8faad09247038f6d2e  guix-build-7b80c36c6f27/output/powerpc64-linux-gnu/bitcoin-7b80c36c6f27-powerpc64-linux-gnu.tar.gz
9fc17890ecaa6307a541ff4160e06f0f933b7ca2ec7e12865c1824e06bbe52c7  guix-build-7b80c36c6f27/output/powerpc64le-linux-gnu/SHA256SUMS.part
75f356a5ae0f1ab2ceea14bc7e1cd2387cdf6290abb7ef93e2060e77dd613476  guix-build-7b80c36c6f27/output/powerpc64le-linux-gnu/bitcoin-7b80c36c6f27-powerpc64le-linux-gnu-debug.tar.gz
46b418a61516e1c13227db9a4b37083142b2aeeded4bc0e2e7e9880278c32203  guix-build-7b80c36c6f27/output/powerpc64le-linux-gnu/bitcoin-7b80c36c6f27-powerpc64le-linux-gnu.tar.gz
048df4422cbb6624b30c15c0baae789927ef39e8ce2a8a270893d67bafa96bf3  guix-build-7b80c36c6f27/output/riscv64-linux-gnu/SHA256SUMS.part
4c2d1d4909ccc434f61b018f9971ab6f059188c58f865bcdba7260be45641c2e  guix-build-7b80c36c6f27/output/riscv64-linux-gnu/bitcoin-7b80c36c6f27-riscv64-linux-gnu-debug.tar.gz
7a5410c52b9108e73bb6f57d3765a431572418a2d4118da6283531bee781c949  guix-build-7b80c36c6f27/output/riscv64-linux-gnu/bitcoin-7b80c36c6f27-riscv64-linux-gnu.tar.gz
4edfdd6fd90ded7fcaf401feef6646d19134903a2596584d64f7aaa8663d1a95  guix-build-7b80c36c6f27/output/x86_64-apple-darwin/SHA256SUMS.part
546a46042ac1e5b95d1cefd006767830ac782ee343a4054a34e32598f582a546  guix-build-7b80c36c6f27/output/x86_64-apple-darwin/bitcoin-7b80c36c6f27-osx-unsigned.dmg
60b68a82c6c24e3b93d78c98d9027838aa04edfe87e5fa9210d86691ec2ef7e9  guix-build-7b80c36c6f27/output/x86_64-apple-darwin/bitcoin-7b80c36c6f27-osx-unsigned.tar.gz
66d2a56a4d7770117e7b976ddd54e3203ae7b49b13e17febe024a5715c9a0d40  guix-build-7b80c36c6f27/output/x86_64-apple-darwin/bitcoin-7b80c36c6f27-osx64.tar.gz
a885ead6819317f82763712b218796fc03d4504887d4d22ead288cd1d5df6382  guix-build-7b80c36c6f27/output/x86_64-linux-gnu/SHA256SUMS.part
44b87f742d2647f07134ad45b6a452d643b00bfcbf2b2527616b884c1acd7358  guix-build-7b80c36c6f27/output/x86_64-linux-gnu/bitcoin-7b80c36c6f27-x86_64-linux-gnu-debug.tar.gz
3a878e0574439df13ad4bbbe7532ac7578b81180530cfddb8863e557c5a1556a  guix-build-7b80c36c6f27/output/x86_64-linux-gnu/bitcoin-7b80c36c6f27-x86_64-linux-gnu.tar.gz
```

MarcoFalke commented:
Tested-only ACK 7b80c36c6f27a03d66d9348325e6ba6cf2cb9b59 (didn't review)

with the steps to reproduce from https://github.com/bitcoin/bitcoin/issues/24314

fanquake commented on configure.ac:
```diff
@@ -1427,8 +1427,21 @@ if test "$use_external_signer" != "no"; then
     *)
-      use_external_signer="yes"
-      AC_DEFINE([ENABLE_EXTERNAL_SIGNER], [1], [Define if external signer support is enabled])
+      AC_MSG_CHECKING([whether Boost.Process can be linked])
```
```suggestion
      AC_MSG_CHECKING([whether Boost.Process can be used])
```

fanquake commented on configure.ac:
```diff
@@ -1427,8 +1427,21 @@ if test "$use_external_signer" != "no"; then
+        if test "$use_external_signer" = "yes"; then
+          AC_MSG_ERROR([External signing is not supported for this Boost version])
+        fi
+        AC_MSG_WARN([External signing is not supported for this Boost version, disabling])
```
I don't think we need this warning.

fanquake reviewed (commented):
Can you rebase this on master?

hebasto commented:
Updated 7b80c36c6f27a03d66d9348325e6ba6cf2cb9b59 -> dec38260d879291fdac52c181c784361b36a5ad5 ([pr24254.01](https://github.com/hebasto/bitcoin/commits/pr24254.01) -> [pr24254.02](https://github.com/hebasto/bitcoin/commits/pr24254.02)):

- rebased on master
- addressed @fanquake's comments

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
d2262c10bc46877d1b2714a208f5accc340a2e9961c3e41e2cb67f424565e837  guix-build-dec38260d879/output/aarch64-linux-gnu/SHA256SUMS.part
94126632e8efb920739ffd40b2a6df2b53bb5f5db1e94d129aa49d6c997c4e91  guix-build-dec38260d879/output/aarch64-linux-gnu/bitcoin-dec38260d879-aarch64-linux-gnu-debug.tar.gz
57d283da672a032dd4306762df41d2d0a3bab1580d30499f7a6c6d8addbf25ad  guix-build-dec38260d879/output/aarch64-linux-gnu/bitcoin-dec38260d879-aarch64-linux-gnu.tar.gz
fca55545088c4eaea6a96026ade5686908c39baebd18420787126ed87c2e6aa4  guix-build-dec38260d879/output/arm-linux-gnueabihf/SHA256SUMS.part
e4a0022212b191b06704003a55b827b84778a3a394bd344cf09e566402899e1f  guix-build-dec38260d879/output/arm-linux-gnueabihf/bitcoin-dec38260d879-arm-linux-gnueabihf-debug.tar.gz
83611b8a6b03c3a099a61da0af11f657939c57ec358b020ec41f863ff45524d4  guix-build-dec38260d879/output/arm-linux-gnueabihf/bitcoin-dec38260d879-arm-linux-gnueabihf.tar.gz
680764b961a13996433b6bf385406c189b1fb468a7810c8c2434cd830e5fc812  guix-build-dec38260d879/output/arm64-apple-darwin/SHA256SUMS.part
a73b8561d0440a89f1682d76e9dbab3fb8f85c67c21bf603afc1d8795cf167b4  guix-build-dec38260d879/output/arm64-apple-darwin/bitcoin-dec38260d879-arm64-apple-darwin.tar.gz
6fb53ac8d07329888f2ece2d049e21eb9123ffd586cbcf92a37c982a9ab4c00d  guix-build-dec38260d879/output/arm64-apple-darwin/bitcoin-dec38260d879-osx-unsigned.dmg
69ab1a7b96b1b07db4c85cd7716841a7c8a85a934b3a0ffa7bd1b56e5e55dd50  guix-build-dec38260d879/output/arm64-apple-darwin/bitcoin-dec38260d879-osx-unsigned.tar.gz
890ddc0b38217302446bbaf231592b4b1636a081e8cd000fe3f2b5cba0972990  guix-build-dec38260d879/output/dist-archive/bitcoin-dec38260d879.tar.gz
50dd6f73da58e026f8bf3fd2cf468b7da70d26db696eb31a442782fd5bd957d9  guix-build-dec38260d879/output/powerpc64-linux-gnu/SHA256SUMS.part
a34130b27440f4867db3c859690e0b2e39e526b97bf9fb66a8d6399394397f47  guix-build-dec38260d879/output/powerpc64-linux-gnu/bitcoin-dec38260d879-powerpc64-linux-gnu-debug.tar.gz
cb17a55906557fc1e78410cf13c36bfdd3f6bdcb0efd9103e8a6a7269420066a  guix-build-dec38260d879/output/powerpc64-linux-gnu/bitcoin-dec38260d879-powerpc64-linux-gnu.tar.gz
d826deb430ea012cd7fee18669e2b741c56a1ee3c34c1441b123b0b55d32dbd2  guix-build-dec38260d879/output/powerpc64le-linux-gnu/SHA256SUMS.part
c72b69a056cee2d0420443977d723c41c7a6fd807de21166aaf34971dd7dde3a  guix-build-dec38260d879/output/powerpc64le-linux-gnu/bitcoin-dec38260d879-powerpc64le-linux-gnu-debug.tar.gz
238a8f8edbd887a0061abd741dd53d94ef729df82fcdc3a1d93e7e598259c4b4  guix-build-dec38260d879/output/powerpc64le-linux-gnu/bitcoin-dec38260d879-powerpc64le-linux-gnu.tar.gz
94a36fc0b05e27ed1d79c7a4323188bd620be62db6761baf3b40d3fd75da64ca  guix-build-dec38260d879/output/riscv64-linux-gnu/SHA256SUMS.part
d9fb148b93bc285e6c9a4c31999b5e40e1cf19338bc05a54723c5b49c9691e43  guix-build-dec38260d879/output/riscv64-linux-gnu/bitcoin-dec38260d879-riscv64-linux-gnu-debug.tar.gz
5b15700377bcfa2adf4627de7b3686046181b82847b102c22da9cd2facce909d  guix-build-dec38260d879/output/riscv64-linux-gnu/bitcoin-dec38260d879-riscv64-linux-gnu.tar.gz
69ea408c36af14d7ddb81c6b461c4ce9bc14cec2ffa321c92e0d49600cf6f134  guix-build-dec38260d879/output/x86_64-apple-darwin/SHA256SUMS.part
e64d822c2db52ca56af467d05611bbc23f4ad1ed441535a03dd6c0e0151dd652  guix-build-dec38260d879/output/x86_64-apple-darwin/bitcoin-dec38260d879-osx-unsigned.dmg
8b5cf3852b821587d0c91f6d5747dc73d3bc6ba5ba0c7dceeb30147315b4ece9  guix-build-dec38260d879/output/x86_64-apple-darwin/bitcoin-dec38260d879-osx-unsigned.tar.gz
c119d14902d3da34099cad63e0ac4305d5a2aae43cd942ffdf363d1d1fe55393  guix-build-dec38260d879/output/x86_64-apple-darwin/bitcoin-dec38260d879-osx64.tar.gz
25e004b9a7809acc05717e4105f462cbbb6068bfe2ceb22f34d2992eab789066  guix-build-dec38260d879/output/x86_64-linux-gnu/SHA256SUMS.part
dae6df4820f1292f0cde621152eb51449aec3cf5811f6f78967297f115b44308  guix-build-dec38260d879/output/x86_64-linux-gnu/bitcoin-dec38260d879-x86_64-linux-gnu-debug.tar.gz
e05cf015190a346df227f2b2d77c1bb352c307d34407620430be21d457e0ef97  guix-build-dec38260d879/output/x86_64-linux-gnu/bitcoin-dec38260d879-x86_64-linux-gnu.tar.gz
```

prusnak reviewed (approved):
Approach ACK dec38260d879291fdac52c181c784361b36a5ad5

hebasto commented:
Updated dec38260d879291fdac52c181c784361b36a5ad5 -> abc057c6030b2a0ddab46835a7801054da677781 ([pr24254.02](https://github.com/hebasto/bitcoin/commits/pr24254.02) -> [pr24254.03](https://github.com/hebasto/bitcoin/commits/pr24254.03), [diff](https://github.com/hebasto/bitcoin/compare/pr24254.02..pr24254.03)):

- improved commit message (thanks to @fanquake)
- no changes in the code

Also the PR description has been updated.

fanquake reviewed (approved):
ACK abc057c6030b2a0ddab46835a7801054da677781

Guix Build:
```bash
bash-5.1# find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
c9113a6a8433169232d63ac52750250467dddde1896465b791047879c9443667  guix-build-abc057c6030b/output/aarch64-linux-gnu/SHA256SUMS.part
ef301fac18cbb1578cfc9d7bc641867ab3dc70c3a18aaa69bd3894a383608e3a  guix-build-abc057c6030b/output/aarch64-linux-gnu/bitcoin-abc057c6030b-aarch64-linux-gnu-debug.tar.gz
bfcefa06e485bd65c96c742591dc3647e6609223ed48bd25a1ac671122eb1fb3  guix-build-abc057c6030b/output/aarch64-linux-gnu/bitcoin-abc057c6030b-aarch64-linux-gnu.tar.gz
b248287deaa79473ced7c315593621640c5f5cf35aa41a7392d5cc79a96895cc  guix-build-abc057c6030b/output/arm-linux-gnueabihf/SHA256SUMS.part
3b46d7640aa8bf6cfec2cc94e220a2de13416558afb63af67ff7172094e4e6c1  guix-build-abc057c6030b/output/arm-linux-gnueabihf/bitcoin-abc057c6030b-arm-linux-gnueabihf-debug.tar.gz
626d91e86e37e98b46cc0231889077c72b41990ffaf08e6985b9f05e428a5475  guix-build-abc057c6030b/output/arm-linux-gnueabihf/bitcoin-abc057c6030b-arm-linux-gnueabihf.tar.gz
fa30549e1a0d1b32355685a1821536673aec765dd1c381aa734435ac81455289  guix-build-abc057c6030b/output/arm64-apple-darwin/SHA256SUMS.part
339d73eb56253e852ef0941334d6314e1bd2f9d34f0a6870e919dd80738c6618  guix-build-abc057c6030b/output/arm64-apple-darwin/bitcoin-abc057c6030b-arm64-apple-darwin.tar.gz
7de1684da96bb4508f06eb24f198b9ce51c05ddf346ea7d7b34ba11a58e36f3f  guix-build-abc057c6030b/output/arm64-apple-darwin/bitcoin-abc057c6030b-osx-unsigned.dmg
06668ed7466b8a71fb5da7227ed110358441dc199bb2ac8fdaeaceac5740bd0e  guix-build-abc057c6030b/output/arm64-apple-darwin/bitcoin-abc057c6030b-osx-unsigned.tar.gz
9cf414febbb6f45d127472b702da28dbeb2c4a95bcf0a48c407c93add2a1ca87  guix-build-abc057c6030b/output/dist-archive/bitcoin-abc057c6030b.tar.gz
539ee17894c799b4fd0b97683e640eab7d6414c728d688725e467ce182b0cc04  guix-build-abc057c6030b/output/powerpc64-linux-gnu/SHA256SUMS.part
72ef9a1b6969c1d6147dfd54c320754b66fbc567b874f7e0ff0620ce59e0c3da  guix-build-abc057c6030b/output/powerpc64-linux-gnu/bitcoin-abc057c6030b-powerpc64-linux-gnu-debug.tar.gz
58a29daa3441c612a5ae5635952f77d4854480ae9c372fe8b525370d5938c138  guix-build-abc057c6030b/output/powerpc64-linux-gnu/bitcoin-abc057c6030b-powerpc64-linux-gnu.tar.gz
e2f0e7f046b5bed003827388a57f38c24a173a1f6b934ee025667f577493070e  guix-build-abc057c6030b/output/powerpc64le-linux-gnu/SHA256SUMS.part
0c9b5e1f5f327510510830417a5f11a2d8151be1d573f30ad7d2396739a4fdd0  guix-build-abc057c6030b/output/powerpc64le-linux-gnu/bitcoin-abc057c6030b-powerpc64le-linux-gnu-debug.tar.gz
0d5f62e66e39587c3e7b4cf553e386ab726d180473dad8dde98443330cae86b2  guix-build-abc057c6030b/output/powerpc64le-linux-gnu/bitcoin-abc057c6030b-powerpc64le-linux-gnu.tar.gz
c2f9de02e08d55131c50f5775dc4eab231a96f9bedc9fd7ae3fb245109ad4747  guix-build-abc057c6030b/output/riscv64-linux-gnu/SHA256SUMS.part
e917d5f8cad66e65f4d872a677f158b5f432c03cead6ad3382a4500c565765e8  guix-build-abc057c6030b/output/riscv64-linux-gnu/bitcoin-abc057c6030b-riscv64-linux-gnu-debug.tar.gz
45a86f5ae68cb7de3086c1b1bdc5cce21f668ddec65b5869a60f2943d6e309f6  guix-build-abc057c6030b/output/riscv64-linux-gnu/bitcoin-abc057c6030b-riscv64-linux-gnu.tar.gz
327881fb491cf0694a22c3b272e63c31b91035849a904d192da23b2b1a861a62  guix-build-abc057c6030b/output/x86_64-apple-darwin/SHA256SUMS.part
f4b78503ba7548ebd4e4981cd173f16ddb27e1dcd69a3deb9e01f3e986552f9d  guix-build-abc057c6030b/output/x86_64-apple-darwin/bitcoin-abc057c6030b-osx-unsigned.dmg
b8c640ad53e653bd9e0863a4c03e4d9d29be31e1454809e940bd06670d4e96a0  guix-build-abc057c6030b/output/x86_64-apple-darwin/bitcoin-abc057c6030b-osx-unsigned.tar.gz
c6a810c10d87596482cec27cfe99881f9e7c144e8decb6c2f8ba4bd819d6fc90  guix-build-abc057c6030b/output/x86_64-apple-darwin/bitcoin-abc057c6030b-osx64.tar.gz
1192e6fc933389d29583030082d138c88e79d4fce0ec9c98606adab30a12ae20  guix-build-abc057c6030b/output/x86_64-linux-gnu/SHA256SUMS.part
e4045d834aad9ff457e02c5f74032221c09d2216ea7cb2b1d92911be4be2cbdc  guix-build-abc057c6030b/output/x86_64-linux-gnu/bitcoin-abc057c6030b-x86_64-linux-gnu-debug.tar.gz
1586330dbcbe44d251669059f7acb8289c6372cd2d848164f2ad77c0507c2010  guix-build-abc057c6030b/output/x86_64-linux-gnu/bitcoin-abc057c6030b-x86_64-linux-gnu.tar.gz
```
